src/DynamoCore/Core/CustomNodeManager.cs

```diff
@@ -39,15 +40,21 @@ public class CustomNodeManager : LogSourceBase, ICustomNodeSource, ICustomNodeMa
         /// <param name="nodeFactory">NodeFactory</param>
         /// <param name="migrationManager">MigrationManager</param>
         /// <param name="libraryServices">LibraryServices</param>
-        public CustomNodeManager(NodeFactory nodeFactory, MigrationManager migrationManager, LibraryServices libraryServices)
+        /// <param name="scheduler">DynamoScheduler</param>
+        public CustomNodeManager(NodeFactory nodeFactory, MigrationManager migrationManager, LibraryServices libraryServices, DynamoScheduler scheduler)
         {
             this.nodeFactory = nodeFactory;
             this.migrationManager = migrationManager;
             this.libraryServices = libraryServices;
+            this.scheduler = scheduler;
         }
 
         #region Fields and properties
 
+        private DynamoScheduler scheduler;
+
+        public EngineController EngineController { get; set; }
```

Comment:
@QilongTang I will fix this by just getting more robust about null checks until we figure out the right logger implementation - there are reasons to avoid a singleton in the initial design, I think, but we might want to reevaluate those. I should be checking null anyway because `Save` makes engine optional.

Reply:
You can reevaluate these points, checking `fromJson` function at `WorkspaceModel`. The constructor of WorkspaceModel also receives Engine object.